src/fabric/executor/multi_threaded/executor.rs

```diff
@@ -0,0 +1,288 @@
+//! A multithreaded executor underlying the fabric
+//!
+//! The executor receives IDs of operations that are ready for execution,
+//! executes them, and places the result back into the fabric for further
+//! executions
+
+use std::{collections::HashMap, sync::Arc};
+
+use ark_ec::CurveGroup;
+use crossbeam::queue::SegQueue;
+use kanal::Sender as KanalSender;
+use rayon::{ThreadPool, ThreadPoolBuilder};
+use tracing::log;
+
+use crate::{
+    fabric::{
+        executor::{buffer::GrowableBuffer, ExecutorJobQueue, ExecutorMessage},
+        result::{ResultWaiter, ERR_RESULT_BUFFER_POISONED},
+        OpResult, Operation, OperationId, OperationType,
+    },
+    network::NetworkOutbound,
+    ResultId,
+};
+
+use super::result_buffer::{ParallelResultBuffer, ResultMask};
+
+/// The executor is responsible for executing operation that are ready for
+/// execution, either passed explicitly by the fabric or as a result of a
+/// dependency being satisfied
+pub struct ParallelExecutor<C: CurveGroup> {
+    /// The job queue for the executor
+    job_queue: Arc<SegQueue<ExecutorMessage<C>>>,
+    /// The operation buffer, stores in-flight operations
+    operations: GrowableBuffer<Operation<C>>,
+    /// The dependency map; maps in-flight results to operations that are
+    /// waiting for them
+    dependencies: GrowableBuffer<Vec<ResultId>>,
+    /// The completed results of operations
+    results: ParallelResultBuffer<C>,
+    /// A bit vector representing which results are available
+    ///
+    /// We use this to avoid contention and races on the `results` buffer while
+    /// the coordinator is computing in-flight args
+    ready_mask: ResultMask,
```

Comment:
note to self: does this need to be concurrent-safe?

Reply:
This one does not need to be, only the coordinator has access to it. It's basically used to avoid coordinator contention with the result queue.